src/Tactic/Rewrite.agda

```diff
@@ -0,0 +1,222 @@
+open import Reflection.Name                 as Name
+open import Reflection.Term                 as Term
+
+open import Reflection.TypeChecking.Monad.Syntax
+
+----------------------------------------------------------------------
+-- Utilities
+----------------------------------------------------------------------
+
+private
+  -- Descend past a variable.
+  var-descend : ℕ → ℕ → ℕ
+  var-descend ϕ x = if ϕ Nat.≤ᵇ x then suc x else x
+
+  -- Descend a variable underneath pattern variables.
+  pattern-descend : ℕ → Pattern → (Pattern × ℕ)
+  patterns-descend : ℕ → Args Pattern → (Args Pattern × ℕ)
+
+  pattern-descend ϕ (con c ps) =
+    let (ps' , ϕ') = patterns-descend ϕ ps
+    in (con c ps' , ϕ')
+  pattern-descend ϕ (dot t)    = (dot t) , ϕ
+  pattern-descend ϕ (var x)    = var (var-descend ϕ x) , suc ϕ
+  pattern-descend ϕ (lit l)    = (lit l) , ϕ
+  pattern-descend ϕ (proj f)   = (proj f) , ϕ
+  pattern-descend ϕ (absurd x) = absurd (var-descend ϕ x) , suc ϕ
+
+  patterns-descend ϕ ((arg i p) ∷ ps) =
+    let (p' , ϕ') = pattern-descend ϕ p
+        (ps' , ϕ'') = patterns-descend ϕ' ps
+    in (arg i p ∷ ps' , ϕ'')
+  patterns-descend ϕ []       =
+    [] , ϕ
+
+  -- Helper for constructing applications of 'cong'
+  `cong : Term → Term → Term
+  `cong f eq = def (quote cong) (4 ⋯⟅∷⟆ vArg (lam visible (abs "ϕ" f)) ∷ 2 ⋯⟅∷⟆ vArg eq ∷ [])
+
+  -- Construct an error when the goal is not 'x ≡ y' for some 'x' and 'y'.
+  not-equality-error : ∀ {A : Set} Term → TC A
+  not-equality-error goal = typeError (strErr "Cannot rewrite a goal that is not equality: " ∷ termErr goal ∷ [])
+
+  -- Extract out both endpoints of an equality type.
+  endpoints : Term → TC (Term × Term)
+  endpoints goal@(def x (lvl ∷ tp ∷ (arg _ e0) ∷ (arg _ e1) ∷ [])) =
+    if x Name.≡ᵇ (quote _≡_) then return (e0 , e1) else not-equality-error goal 
+  endpoints (meta m args) = blockOnMeta m
+  endpoints goal = not-equality-error goal 
+
+----------------------------------------------------------------------
+-- Anti-Unification
+--
+-- The core idea of the tactic is that we can compute the input
+-- to 'cong' by syntactically anti-unifying both sides of the
+-- equality, and then using that to construct a lambda
+-- where all the differences are replaced by the lambda-abstracted
+-- variable.
+--
+-- For instance, the two terms 'suc (m + (m + 0)) + (m + 0)' and
+-- 'suc (m + m) + (m + 0)' would anti unify to 'suc (m + _) + (m + 0)'
+-- which we can then use to construct the lambda 'λ ϕ → suc (m + ϕ) + (m + 0)'.
+----------------------------------------------------------------------
+
+private
+  anti-unify : ℕ → Term → Term → Term
+  anti-unify-args : ℕ → Args Term → Args Term → Maybe (Args Term)
+  anti-unify-clauses : ℕ → Clauses → Clauses → Maybe Clauses
+  anti-unify-clause : ℕ → Clause → Clause → Maybe Clause
+  
+  anti-unify ϕ (var x args) (var y args') with x Nat.≡ᵇ y | anti-unify-args ϕ args args'
+  ... | _     | nothing    = var ϕ []
```

Comment:
Eek, that's a bit far even for me :fearful:

Reply:
There's a reason I left it as a comment rather than a 'must change'.

The main point of (experimentally) changing the code to be like that would be to see the various patterns in the code appear more clearly. Right now, the code looks clear, but the patterns are not so easy to see.

To be clear: I'd recommend @TOTBWF experiment on his own branch with that, to see what emerges, if anything. But I'm not seriously suggesting changing it that way.

Reply:
Makes sense. Sorry, didn't mean to come across as judgy!

Reply:
Hey, your job title could be 'judger in chief' - in a good way! Your current role is to be final arbiter of the style of stdlib. You're doing a fine job of it, IMHO.